File: AdisyonUygulamaJavaXMLsiz/app/src/main/java/com/example/adisyonuygulamajava/viewmodel/MasaDetayViewModel.java
package com.example.adisyonuygulamajava.viewmodel;

import android.util.Log;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.adisyonuygulamajava.model.Kategori;
import com.example.adisyonuygulamajava.model.Masa;
import com.example.adisyonuygulamajava.model.MasaUrun;
import com.example.adisyonuygulamajava.model.Urun;
import com.example.adisyonuygulamajava.services.retrofit.AdisyonServisDao;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class MasaDetayViewModel extends ViewModel {

    private final int masaId;
    private final AdisyonServisDao dao = new AdisyonServisDao();
    private final ExecutorService executor = Executors.newSingleThreadExecutor();

    private final MutableLiveData<List<MasaUrun>> _urunler = new MutableLiveData<>();
    private final MutableLiveData<List<Urun>> _tumUrunler = new MutableLiveData<>();
    private final MutableLiveData<List<Kategori>> _kategoriler = new MutableLiveData<>();
    private final MutableLiveData<Float> _toplamFiyat = new MutableLiveData<>();
    private final MutableLiveData<Boolean> _isLoading = new MutableLiveData<>();
    private final MutableLiveData<Masa> _masa = new MutableLiveData<>();
    public final MutableLiveData<Boolean> odemeTamamlandi = new MutableLiveData<>(false);

    public MasaDetayViewModel(int masaId) {
        this.masaId = masaId;
    }

    public LiveData<List<MasaUrun>> getUrunler() {
        return _urunler;
    }

    public LiveData<List<Urun>> getTumUrunler() {
        return _tumUrunler;
    }

    public LiveData<List<Kategori>> getKategoriler() {
        return _kategoriler;
    }

    public LiveData<Float> getToplamFiyat() {
        return _toplamFiyat;
    }

    public LiveData<Boolean> getIsLoading() {
        return _isLoading;
    }

    public LiveData<Masa> getMasa() {
        return _masa;
    }

    public void yukleTumVeriler() {
        executor.execute(() -> {
            _isLoading.postValue(true);
            try {
                Masa masa = dao.masaGetir(masaId);
                _masa.postValue(masa);

                List<MasaUrun> masaUrunleri = dao.masaUrunleriniGetir(masaId);
                List<Urun> urunler = dao.urunleriGetir();
                List<Kategori> kategoriler = dao.kategorileriGetir();

                List<Urun> guncellenmisUrunler = new ArrayList<>();
                for (Urun urun : urunler) {
                    int adet = 0;
                    for (MasaUrun mu : masaUrunleri) {
                        if (mu.getUrun_id() == urun.getId()) {
                            adet = mu.getAdet();
                            break;
                        }
                    }
                    urun.setUrun_adet(adet);
                    guncellenmisUrunler.add(urun);
                }

                float toplam = 0f;
                for (MasaUrun mu : masaUrunleri) {
                    toplam += mu.getToplam_fiyat();
                }

                _urunler.postValue(masaUrunleri);
                _tumUrunler.postValue(guncellenmisUrunler);
                _kategoriler.postValue(kategoriler);
                _toplamFiyat.postValue(toplam);

            } catch (Exception e) {
                Log.e("VM", "Veri yüklenemedi: " + e.getLocalizedMessage());
            } finally {
                _isLoading.postValue(false);
            }
        });
    }

    public void urunEkle(int urunId, int adet) {
        executor.execute(() -> {
            try {
                dao.urunEkle(masaId, urunId, adet);
                List<MasaUrun> masaUrunleri = dao.masaUrunleriniGetir(masaId);
                float toplam = 0f;
                for (MasaUrun mu : masaUrunleri) {
                    toplam += mu.getToplam_fiyat();
                }

                _urunler.postValue(masaUrunleri);
                _toplamFiyat.postValue(toplam);
                yukleTumVeriler();
            } catch (Exception e) {
                Log.e("VM", "Ürün ekleme hatası: " + e.getLocalizedMessage());
            }
        });
    }

    public void urunCikar(int urunId) {
        executor.execute(() -> {
            try {
                dao.urunCikar(masaId, urunId);

                List<Urun> urunler = _tumUrunler.getValue();
                if (urunler != null) {
                    List<Urun> guncellenmis = new ArrayList<>();
                    for (Urun u : urunler) {
                        if (u.getId() == urunId && u.getUrun_adet() > 0) {
                            u.setUrun_adet(u.getUrun_adet() - 1);
                        }
                        guncellenmis.add(u);
                    }
                    _tumUrunler.postValue(guncellenmis);
                }

                List<MasaUrun> masaUrunleri = dao.masaUrunleriniGetir(masaId);
                float toplam = 0f;
                for (MasaUrun mu : masaUrunleri) {
                    toplam += mu.getToplam_fiyat();
                }

                _urunler.postValue(masaUrunleri);
                _toplamFiyat.postValue(toplam);
                yukleTumVeriler();
            } catch (Exception e) {
                Log.e("VM", "Ürün çıkarma hatası: " + e.getLocalizedMessage());
            }
        });
    }

    public void odemeAl(Runnable onSuccess) {
        executor.execute(() -> {
            try {
                dao.masaOdemeYap(masaId);

                List<Urun> urunler = _tumUrunler.getValue();
                if (urunler != null) {
                    for (Urun u : urunler) {
                        u.setUrun_adet(0);
                    }
                    _tumUrunler.postValue(urunler);
                }

                _toplamFiyat.postValue(0f);

                // Ödeme tamamlandı
                odemeTamamlandi.postValue(true);

                if (onSuccess != null) onSuccess.run();

            } catch (Exception e) {
                Log.e("VM", "Ödeme alma hatası: " + e.getLocalizedMessage());
            }
        });
    }

}

File: AdisyonUygulamaKotlinXMLsiz/app/src/main/java/com/example/adisyonuygulamakotlin/viewmodel/MasaDetayViewModel.kt
package com.example.adisyonuygulamakotlin.viewmodel

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.adisyonuygulamakotlin.model.Kategori
import com.example.adisyonuygulamakotlin.model.Masa
import com.example.adisyonuygulamakotlin.model.MasaUrun
import com.example.adisyonuygulamakotlin.model.Urun
import com.example.adisyonuygulamakotlin.services.retrofit.AdisyonServisDao
import kotlinx.coroutines.launch


class MasaDetayViewModel(private val masaId: Int) : ViewModel() {

    private val dao = AdisyonServisDao()

    private val _urunler = MutableLiveData<List<MasaUrun>>()             // masaya ait ürünler
    val urunler: LiveData<List<MasaUrun>> = _urunler

    private val _tumUrunler = MutableLiveData<List<Urun>>()              // tüm ürünler
    val tumUrunler: LiveData<List<Urun>> = _tumUrunler

    private val _kategoriler = MutableLiveData<List<Kategori>>()         // kategoriler
    val kategoriler: LiveData<List<Kategori>> = _kategoriler

    private val _toplamFiyat = MutableLiveData<Float>()                  // toplam fiyat
    val toplamFiyat: LiveData<Float> = _toplamFiyat

    private val _isLoading = MutableLiveData<Boolean>()
    val isLoading: LiveData<Boolean> = _isLoading

    private val _masa = MutableLiveData<Masa>()
    val masa: LiveData<Masa> = _masa


    fun yukleTumVeriler() {
        viewModelScope.launch {
            _isLoading.value = true

            try {
                val masa = dao.masaGetir(masaId)
                _masa.value = masa

                val masaUrunleri = dao.masaUrunleriniGetir(masaId)
                val urunler = dao.urunleriGetir()
                val kategoriler = dao.kategorileriGetir()

                // Masadaki ürün adetlerini Urun listesine eşleştir
                val guncellenmisUrunler = urunler.map { urun ->
                    val eslesenMasaUrun = masaUrunleri.find { it.urun_id == urun.id }
                    if (eslesenMasaUrun != null) {
                        urun.copy(urun_adet = eslesenMasaUrun.adet)
                    } else {
                        urun.copy(urun_adet = 0)
                    }
                }

                // LiveData güncellemeleri
                _urunler.value = masaUrunleri
                _tumUrunler.value = guncellenmisUrunler
                _kategoriler.value = kategoriler
                _toplamFiyat.value = masaUrunleri.sumOf { it.toplam_fiyat.toDouble() }.toFloat()

            } catch (e: Exception) {
                Log.e("VM", "Veri yüklenemedi: ${e.localizedMessage}")
            } finally {
                _isLoading.value = false
            }
        }
    }


    fun urunEkle(urunId: Int, adet: Int = 1) {
        viewModelScope.launch {
            try {
                dao.urunEkle(masaId, urunId, adet)
                // Masa ürünlerini ve toplam fiyatı güncelle
                val masaUrunleri = dao.masaUrunleriniGetir(masaId)
                _urunler.postValue(masaUrunleri)
                _toplamFiyat.postValue(masaUrunleri.sumOf { it.toplam_fiyat.toDouble() }.toFloat())
                yukleTumVeriler()
2

            } catch (e: Exception) {
                Log.e("VM", "Ürün ekleme hatası: ${e.localizedMessage}")
            }
        }
    }


    fun urunCikar(urunId: Int) {
        viewModelScope.launch {
            dao.urunCikar(masaId, urunId)

            // tumUrunler listesindeki ilgili ürünün adetini azalt
            _tumUrunler.value = _tumUrunler.value?.map {
                if (it.id == urunId && it.urun_adet > 0) {
                    it.copy(urun_adet = it.urun_adet - 1)
                } else it
            }

            // masaya ait ürünleri ve toplam fiyatı güncelle
            val masaUrunleri = dao.masaUrunleriniGetir(masaId)
            _urunler.postValue(masaUrunleri)
            _toplamFiyat.postValue(masaUrunleri.sumOf { it.toplam_fiyat.toDouble() }.toFloat())
            yukleTumVeriler()
        }
    }

    val odemeTamamlandi = MutableLiveData<Boolean>(false)


    fun odemeAl(onSuccess: () -> Unit) {
        viewModelScope.launch {
            dao.masaOdemeYap(masaId)

            // tumUrunler listesindeki tüm adetleri sıfırla
            _tumUrunler.value = _tumUrunler.value?.map {
                it.copy(urun_adet = 0)
            }
            _toplamFiyat.value = 0f
            odemeTamamlandi.postValue(true)
            onSuccess()
        }
    }

}
